test/auth.test.ts

```diff
@@ -0,0 +1,356 @@
+        }),
+      });
+    });
+
+    afterEach(() => {
+      auth = undefined as any;
+    });
+
+    it('rejects if already logging in', () => {
+      auth.isLoggingIn = true;
+      return expect(auth.login()).rejects.toBeDefined();
+    });
+
+    it('shows console error if no client id set', () => {
+      const consoleSpy = jest.spyOn(console, 'error');
+      consoleSpy.mockImplementation(() => {});
+      const codeGenSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      auth = new Auth({} as any);
+      auth.login();
+      expect(consoleSpy).toHaveBeenCalled();
+      expect(codeGenSpy).toHaveBeenCalledTimes(0);
+
+      consoleSpy.mockRestore();
+      codeGenSpy.mockRestore();
+    });
+
+    it('opens a window to the wally api login page with clientId and state', () => {
+      const state = 'mississippi';
+      const storageSpy = jest.spyOn(Storage.prototype, 'setItem');
+
+      jest
+        .spyOn(Auth.prototype as any, 'generateStateCode')
+        .mockImplementation(() => state);
+
+      auth.login();
+
+      expect(storageSpy).toHaveBeenCalledWith(
+        `wally:${clientId}:state:token`,
+        state
+      );
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}`,
+        '_blank'
+      );
+
+      storageSpy.mockRestore();
+    });
+
+    it('passes redirectUrl and email when set', () => {
+      const state = 'mississippi';
+      const email = 'joe@smith.net';
+      const redirectURL = 'http://salmonofcapistrano.com';
+
+      const genCodeSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      genCodeSpy.mockImplementation(() => state);
+
+      auth = new Auth({
+        clientId,
+        host,
+        redirectURL,
+        messenger: new Messenger({}),
+      } as any);
+      auth.login(email);
+
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}&redirectUrl=${encodeURIComponent(
+          redirectURL
+        )}&email=${encodeURIComponent(email)}`,
+        '_blank'
+      );
+      genCodeSpy.mockRestore();
+    });
+
+    describe('returned promise', () => {
+      it('attaches message listeners', () => {
+        const emitterSpy = jest.spyOn(Messenger.prototype, 'addListener');
+        const workerSpy = jest.spyOn(Messenger.prototype, 'onWorkerMessage');
+        auth.login();
+        expect(emitterSpy).toHaveBeenCalled();
+        expect(workerSpy).toHaveBeenCalledTimes(2);
+        expect(workerSpy.mock.calls[0][0]).toBe(WorkerMessage.LOGIN_SUCCESS);
+        expect(workerSpy.mock.calls[1][0]).toBe(WorkerMessage.LOGIN_FAILURE);
+
+        emitterSpy.mockRestore();
+        workerSpy.mockRestore();
+      });
+
+      it('resolves on the accounts changed emitter message', async () => {
+        setTimeout(() => {
+          auth['messenger'].emit(EmitterMessage.ACCOUNTS_CHANGED, '0x1234');
+        }, 0);
+
+        await auth.login();
```

Comment:
Q: Is there an assertion missing here?

Reply:
well, the promise returned by `login()` doesn't resolve with any value, so the internet said a simple `await` was basically the same as asserting a resolution. But I agree, it definitely looks a little funky, i'll make it more explicit